%% BioMed_Central_Tex_Template_v1.06
%%                                      %
%  bmc_article.tex            ver: 1.06 %
%                                       %

%%IMPORTANT: do not delete the first line of this template
%%It must be present to enable the BMC Submission system to
%%recognise this template!!

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                     %%
%%  LaTeX template for BioMed Central  %%
%%     journal article submissions     %%
%%                                     %%
%%          <8 June 2012>              %%
%%                                     %%
%%                                     %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                                                 %%
%% For instructions on how to fill out this Tex template           %%
%% document please refer to Readme.html and the instructions for   %%
%% authors page on the biomed central website                      %%
%% http://www.biomedcentral.com/info/authors/                      %%
%%                                                                 %%
%% Please do not use \input{...} to include other tex files.       %%
%% Submit your LaTeX manuscript as one .tex document.              %%
%%                                                                 %%
%% All additional figures and files should be attached             %%
%% separately and not embedded in the \TeX\ document itself.       %%
%%                                                                 %%
%% BioMed Central currently use the MikTex distribution of         %%
%% TeX for Windows) of TeX and LaTeX.  This is available from      %%
%% http://www.miktex.org                                           %%
%%                                                                 %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%% additional documentclass options:
%  [doublespacing]
%  [linenumbers]   - put the line numbers on margins

%%% loading packages, author definitions

%\documentclass[twocolumn]{bmcart}% uncomment this for twocolumn layout and comment line below
\documentclass{bmcart}

%%% Load packages
%\usepackage{amsthm,amsmath}
%\RequirePackage{natbib}
\RequirePackage{hyperref}
\usepackage[utf8]{inputenc} %unicode support
%\usepackage[applemac]{inputenc} %applemac support if unicode package fails
%\usepackage[latin1]{inputenc} %UNIX support if unicode package fails

\usepackage{color}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                             %%
%%  If you wish to display your graphics for   %%
%%  your own use using includegraphic or       %%
%%  includegraphics, then comment out the      %%
%%  following two lines of code.               %%
%%  NB: These line *must* be included when     %%
%%  submitting to BMC.                         %%
%%  All figure files must be submitted as      %%
%%  separate graphics through the BMC          %%
%%  submission process, not included in the    %%
%%  submitted article.                         %%
%%                                             %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\def\includegraphic{}
\def\includegraphics{}



%%% Put your definitions there:
\startlocaldefs
\endlocaldefs


%%% Begin ...
\begin{document}

%%% Start of article front matter
\begin{frontmatter}

\begin{fmbox}
\dochead{Methodology}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the title of your article here     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\title{Wanderer, an interactive viewer to explore DNA methylation and gene expression data in human cancer}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors here                   %%
%%                                          %%
%% Specify information, if available,       %%
%% in the form:                             %%
%%   <key>={<id1>,<id2>}                    %%
%%   <key>=                                 %%
%% Comment or delete the keys which are     %%
%% not used. Repeat \author command as much %%
%% as required.                             %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


\author[
   addressref={aff1,aff2},
   % corref={aff1},
   email={adiez@imppc.org},
   noteref={n1}
]{\inits{A}\fnm{Anna} \snm{Díez-Villanueva}}
\author[
   addressref={aff1,aff2},                   % id's of addresses, e.g. {aff1,aff2}
   % corref={aff1},                       % id of corresponding address, if any
   noteref={n1},                        % id's of article notes, if any
   email={imallona@imppc.org}   % email address
]{\inits{I}\fnm{Izaskun} \snm{Mallona}}
\author[
   addressref={aff1,aff2},
   corref={aff1,aff2},
   % noteref={n2},
   email={map@imppc.org}
]{\inits{MA}\fnm{Miguel A.} \snm{Peinado}}



% \author[
%    addressref={aff1},                   % id's of addresses, e.g. {aff1,aff2}
%    corref={aff1},                       % id of corresponding address, if any
%    noteref={n1},                        % id's of article notes, if any
%    email={jane.e.doe@cambridge.co.uk}   % email address
% ]{\inits{JE}\fnm{Jane E} \snm{Doe}}
% \author[
%    addressref={aff1,aff2},
%    email={john.RS.Smith@cambridge.co.uk}
% ]{\inits{JRS}\fnm{John RS} \snm{Smith}}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter the authors' addresses here        %%
%%                                          %%
%% Repeat \address commands as much as      %%
%% required.                                %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%





\address[id=aff1]{%                           % unique id
  \orgname{Health Research Institute Germans Trias i Pujol (IGTP) }, % university, etc
  % \street{Can Ruti Campus. Ctra. de Can Ruti, camí de les escoles, s/n},                     %
  % \postcode{08916},                                % post or zip code
  % \city{Badalona},                              % city
  % \cny{Spain}                               
  % country
}

\address[id=aff2]{%                           % unique id
  \orgname{Institute of Predictive and Personalized Medicine of Cancer (IMPPC)}, % university, etc
  \street{Can Ruti Campus. Ctra. de Can Ruti, camí de les escoles, s/n},                     %
  \postcode{08916},                                % post or zip code
  \city{Badalona},                              % city
  \cny{Spain}                                    % country
}
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Enter short notes here                   %%
%%                                          %%
%% Short notes will be after addresses      %%
%% on first page.                           %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{artnotes}
%\note{Sample of title note}     % note to the article
\note[id=n1]{Equal contributor} % note, connected to author
% \note[id=n2]{Corresponding author}
\end{artnotes}

\end{fmbox}% comment this for two column layout

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Abstract begins here                 %%
%%                                          %%
%% Please refer to the Instructions for     %%
%% authors on http://www.biomedcentral.com  %%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{abstractbox}


\begin{abstract}

% Wanderer is an intuitive Web tool allowing real time access and visualization of gene expression and DNA methylation profiles from The Cancer Genome Atlas (TCGA). Given a gene query and selection of a TCGA dataset (e.g.: Colon adenocarcinomas), the Web resource provides the expression profile, at the single exon level, and the DNA methylation levels of HumanMethylation450 BeadChip loci inside or in the vicinity of the queried gene. Graphic and table outputs allow the comparison of tumor and normal profiles and the exploration along the genomic locus and across tumor collections. The simple interface provides a straightforward access to TCGA data to experimentalists and clinicians without bioinformatics expertise. Web tool may be accessed at \url{http://maplab.cat/wanderer}.



Background: The Cancer Genome Atlas (TCGA) offers a multilayered view of genomics and epigenomics data of many human cancer types. However, the retrieval of expression and methylation data from TCGA is a time consuming task.\\

Results: Wanderer is an intuitive Web tool allowing real time access and visualization of gene expression and DNA methylation profiles from TCGA. Given a gene query and selection of a TCGA dataset (e.g.: Colon adenocarcinomas), the Web resource provides the expression profile, at the single exon level, and the DNA methylation levels of HumanMethylation450 BeadChip loci inside or in the vicinity of the queried gene. Graphic and table outputs allow the comparison of tumor and normal profiles and the exploration along the genomic locus and across tumor collections.\\

Conclusions: Wanderer offers a simple interface to straightforward access to TCGA data, amenable to experimentalists and clinicians without bioinformatics expertise.  Wanderer may be accessed at \url{http://maplab.cat/wanderer}. 

\end{abstract}


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The keywords begin here                  %%
%%                                          %%
%% Put each keyword in separate \kwd{}.     %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{keyword}
\kwd{data visualization}
\kwd{TCGA}
\kwd{DNA methylation profiling}
\kwd{expression profiling}
\kwd{Web tool}
\end{keyword}

% MSC classifications codes, if any
%\begin{keyword}[class=AMS]
%\kwd[Primary ]{}
%\kwd{}
%\kwd[; secondary ]{}
%\end{keyword}

\end{abstractbox}
%
%\end{fmbox}% uncomment this for twcolumn layout

\end{frontmatter}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% The Main Body begins here                %%
%%                                          %%
%% Please refer to the instructions for     %%
%% authors on:                              %%
%% http://www.biomedcentral.com/info/authors%%
%% and include the section headings         %%
%% accordingly for your article type.       %%
%%                                          %%
%% See the Results and Discussion section   %%
%% for details on how to create sub-sections%%
%%                                          %%
%% use \cite{...} to cite references        %%
%%  \cite{koon} and                         %%
%%  \cite{oreg,khar,zvai,xjon,schn,pond}    %%
%%  \nocite{smith,marg,hunn,advi,koha,mouse}%%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%% start of article main body
% <put your article body there>

%%%%%%%%%%%%%%%%
%% Background %%
%%

\section*{Background}



Cancer includes a group of largely heterogeneous and complex diseases. The development of massive genomic analysis techniques and the coordinated effort of different international consortiums are providing the research community with comprehensive data of tissue samples from thousands of cancer patients. Free access to these data represents an invaluable opportunity for the research community to get insights into the molecular profiles of cancers allowing, for instance, identification of new candidate markers, in silico hypothesis testing or validation of in-house results through an independent series.\\


As a group of largely heterogenous and complex disease, cancer research is benefited by coordinated efforts on data acquisition and analysis. The Cancer Genome Atlas (TCGA) \cite{zhang2011international} is probably the most ambitious of these initiatives offering a multilayered view of genomics and epigenomics data together with clinicopathological information of more than 30 human cancer types. TCGA Data Portal \cite{tcgaportalweb} is the default gateway to retrieve data offering a text file for each processed sample belonging to a dataset subjected to a given platform. Compressed packages for multiple samples are allowed. For instance, it allows downloading 992 files with the clinical data related to samples of the Colon Adenocarcinoma (COAD) dataset, but only provides RNAseq results for 303 of them. It is worth noting that as the number of files and their sizes increase, further processing of this information exceeds the spreadsheet-based analysis capabilities and requires bioinformatic skills (i.e.: programming under the R statistical environment).\\

Cancer research teams working on specific aspects of the disease (e.g.: exploration of candidate biomarkers, deregulation of signaling pathways, or mechanisms of gene regulation, just to name a few) might be more interested in querying specific subsets of the data (e.g.: expression levels of a named gene or the methylation status of its promoter region), rather than obtaining bulky datasets as offered by the the Data Portal. To facilitate the access and exploitation of TCGA data, several tools have been developed allowing from the comparison of gene expression or DNA methylation levels to the exploration of correlations \cite{plass2013mutations,schroeder2013visualizing}.\\


A major contribution of TCGA consortium is illustrated by the opening manuscripts of each one of the considered human cancers \cite{weinstein2013cancer} and other references \cite{tcganetworkweb}. These papers are largely devoted to report the mutational landscapes and several tools have been developed to explore and interpret the significance of these alterations in cancer \cite{international2013computational}. Unlike mutations, the analysis of transcriptomic and epigenetic scenarios in cancer samples represents a much higher level of complexity and important differences among samples may be easily missed if generalized criteria are used to compare data. For instance, gene expression differences may appear as alternative transcripts and be missed in the summarizing values. Moreover, DNA methylation changes may occur outside predefined promoter domains or just affect a subregion.\\ 

The purpose of this work is to develop Wanderer, a very simple and intuitive web tool allowing real time access and visualization of gene expression and DNA methylation profiles from TCGA data using gene targeted queries. Wanderer is addressed to a broad variety of experimentalists and clinicians without deep bioinformatics skills.\\       
  

%%%


% Data analysis on human cancer is a hotspot in basic and clinical research. As a group of largely heterogenous and complex disease, cancer research is benefited by coordinated efforts on data acquisition and analysis. The Cancer Genome Atlas (TCGA) is a result of an international consortium that offers a multilayered view of the genomics and epigenomics of more than 30 human cancer types \cite{weinstein2013cancer}. This enormous data availability opens a wide world of possibilities to the research community. For instance, it allows \textit{in silico} hypothesis testing or validation of in-house results through an independent series.\\


% % The TCGA serves the information in three tiers according to the degree of its bioinformatical processing: raw, lightly and heavily analyzed data (named levels 1, 2 and 3). Data not belonging uniquely to an individual, such as level 3 and de-identified clinical data, is open access; on the other hand, the download of protected data, such as raw sequencing files, requires authentication via the The Cancer Genomics Hub (CGHub)\cite{wilks2014cancer}. Finally, although the open access data tier offers immediate access to the information, some projects restrict the dissemination of the results brought by its use until a first global analysis is published by the consortia.\\

% % The TCGA Data Portal is the gateway to retrieve the TCGA information. Basically, offers a text file for each processed sample belonging to a dataset subjected to a platform; compressed packages for multiple samples are allowed. For instance, it allows downloading 992 files with the clinical data related to samples of the Colon Adenocarcinoma (COAD) dataset, but only provides RNAseq results for 303 of them. It is worth noting that as the number of files and their sizes increase, further processing of this information exceeds the spredsheet-based analysis capabilities and requires some amount of bioinformatics knowledge, (i.e. programming under the R statistical environment). Comprehensive and integrative analysis is indeed dependent on informatics solutions, such as TCGA Assembler \cite{zhu2014tcga} or the cBioPortal\cite{gao2013integrative}.\\

% The TCGA Data Portal is the gateway to retrieve the TCGA information. Basically, offers a text file for each processed sample belonging to a dataset subjected to a platform; compressed packages for multiple samples are allowed. For instance, it allows downloading 992 files with the clinical data related to samples of the Colon Adenocarcinoma (COAD) dataset, but only provides RNAseq results for 303 of them. It is worth noting that as the number of files and their sizes increase, further processing of this information exceeds the spredsheet-based analysis capabilities and requires some amount of bioinformatics knowledge, (i.e. programming under the R statistical environment).\\ %Comprehensive and integrative analysis is indeed dependent on informatics solutions, such as the cBioPortal \cite{gao2013integrative}.\\

% The complex and comprehensive  nature of the generated data represents a privileged opportunity for researcher to get insights into the molecular landscapes of cancers. In fact, the landmark papers reporting the original data of each tumor type may just be considered a modest anticipation of the expected overall outcome. Nevertheless,  wet lab experimentalists have very limited access to these data because sophisticated bioinformatic skills are required to manage and analyze such large datasets. Hence, development of interfaces facilitating the visualization and analysis of the information to non-bioinformaticians is essential to get a broad exploitation of this gigantic effort.\\

% The availability of such a large amount of data is exciting to the cancer research community as it helps to check candidate \textit{loci}. For instance, as some genes show aberrant promoter hypermethylation and therefore silencing during the tumor progression \cite{jones2002fundamental}, experimentalists might query the expression and methylation profiles of their candidate \textit{loci} on tumoral and normal matched tissues. However, the Data Portal does not allow to query individual genomic locations but rather offers bulky, whole genome datasets. To our knowledge, although much effort has been done on facilitating the access to TCGA data, none of the published tools allows an user-friendly interface to solve this kind of queries. We believe that this candidate and exploratory approaches might be particularly benefited by using RNASeq and methylation data. Our response has been the development of Wanderer, a very simple and intuitive web tool allowing real time access and visualization of gene expression and DNA methylation profiles from TCGA data using gene targeted queries.\\


\section*{Results}

{\color{red} \subsection*{Methylation and expression profiling}}

% With this aim we developed a tool, the TCGA Wanderer, that allows to easily retrieve and visualize the regional profiles of groups of samples belonging to a tumor type. 

For any given gene selected by the investigator, the tool provides detailed individual profiles of gene expression, exon by exon, and DNA methylation of all the probes inside or in the vicinity of the gene (Figure \ref{fig:1}). Graphs for normal and tumor samples from any of the available TCGA datasets are readily produced. Summarizing plots and tables are also generated, allowing further data analysis or representation using any software the investigator is used to (e.g.: Excel or any other spreadsheet or graphing tool). In addition, statistical analysis is applied to identify differential gene expression or DNA methylation between normal and tumor samples at either the single exon or the DNA methylation probe level, respectively. The tool allows local navigation and zoom in/out within the region of interest, as well as simple graph customization.\\

{\color{red} To complement these plots, Wanderer produces further different graphical representations depending on whether expression or methylation data are selected}.\\ 

{\color{red}
Regarding expression data, the profiling reflects the expression of each exon measured as Reads per Kilobase per Million mapped reads (RPKM). Wanderer also explores the gene expression between normals and tumors using a summarized-by-gene value. {\color{red} Briefly, this summarization takes into account all the exons of the gene using RSEM values \cite{guo2013large}}. As a result, a boxplot depicting the differences in its overall expression between normal and tumors is produced; a Wilcoxon rank sum test checks whether normal and tumors are statistically equivalent.\\
}

{\color{red}
Wanderer complements the DNA methylation profiling with a tool to overview the influence of DNA methylation on gene expression. This approach generates a correlation analysis of the gene expression value with any of the CpG probes of the region. The user can navigate through upstream, gene body or downstream probes as far as they are in the gene's vicinity.}\\

\subsection*{Plotting versatility}

Graphical outputs are highly dynamic, being rendered on the fly. {\color{red} Once the gene of interest is selected, some parameters can be easily tuned: }

\begin{itemize}
\item The user can zoom in and out either by using a simple slider or entering a custom coordinate range by hand. 
\item A glyph depicting the actual gene location and strand can be added. 
\item Methylation probes belonging to a CpG island can be highlighted in green. 
\item Distance between probes or exons can be selected to be proportional to the actual genomic location or to be uniform; in the case of methylation, this facilitates visualization of CpG-dense regions, such as CpG islands. 
\item In order to ease sample profiling, Wanderer draws line plots by default. In this representation, the scrutinized exons or CpGs belonging to the same sample are linked by lines. The user can enter the number of samples to be represented this way. A random sampling (with a random seed so the result is repeatable) is therefore performed. However, all the data can be taken into account at once ticking the 'Boxplot all the data' checkbox.
\item {\color{red} The user can select the p-value threshold for by-exon or by-CpG normal \textit{vs} tumor comparison.}
\item {\color{red} When profiling DNA methylation data, Wanderer produces a correlation plot of the summarized-by-gene overall expression with a given CpG probe methylation: the user only needs to select one of the CpG probes located inside the gene or in the surrounding area.}
\end{itemize}

{\color{red}
\subsection*{Interplay with other tools}

The tool also provides batch analysis capabilities as well data sharing and linking from other Web resources through an application programming interface (API) named Wanderer\_api \cite{wandererapi}. Wanderer\_api has the same capabilities as Wanderer does but receives the parameters parsing arguments from the URL.\\ 

Wanderer also produces a link to the UCSC Genome Browser's profile of the scrutinized region; and to two TCGA data exploring platforms, the cBioPortal \cite{gao2013integrative} and the All Pairs Regulome Explorer \cite{cancer2012comprehensive}, when available; {\color{red} and the UCSC Genome Browser}. The interplay with cBioPortal facilitates access to mutations, protein and survival data, among others. The Regulome Explorer permits the visualization of mutations, copy number alterations, microRNAs or clinical data and their associations.}


\subsection*{Further capabilities}
Wanderer performs Wilcoxon rank sum test comparisons on normal versus tumor provided there are at least two observations in each group. An independent test is performed for each CpG or exon, being the p-value FDR corrected for multiple testing. Although this is readily reported, we note that addressing the significance on few observations and without inspecting the data distribution might be meaningless. Even though the number of samples to draw as line-plots can be reduced manually, the statistical analysis is performed with the complete dataset.\\





Graphics are readily downloaded {\color{red} in raster (PNG) and vector (PDF) formats}. Probe annotations, statistical analysis, {\color{red} clinical} and raw data for either normal and tumor samples {\color{red} are available} as text files (comma-separated values, CSV) that can be processed using spreadsheet software. {\color{red} Data, analysis and graphics are packaged as a ZIP file with a unique identifier containing the gene name, dataset and the user's timestamp. The ZIP file contains a copy of the user manual including a detailed summary of the outputs and the software and data version.} \\



% Wanderer performs Wilcoxon rank sum test comparisons on normal \textit{vs} tumor when there is at least an observation in each group. An independent test is performed for each CpG or exon, being the p-value FDR corrected for multiple testing. Although this is reported automatedly, we note that addressing the significance on few observations and without inspecting the data distribution might be meaningless. Even though the number of samples to draw as line-plots can be reduced manually, the statistical analysis is performed with the complete dataset.\\

% Graphics can be readily downloaded in either raster (PNG) or vector (PDF) formats. Probe annotations, statistical analysis and raw data for either normal and tumor samples can be easily download as text files (comma-separated values, CSV) that can be processed using spreadsheet software. A link pointing to the UCSC Genome Browser's profile of the scrutinized regions is readily produced.\\

% We provide a simplified flavour of the Web tool named Wanderer\_api (\url{http://maplab.cat/wanderer_api}) that offers batch analysis as well data sharing and linking from other Web resources. Wanderer\_api has the same capabilities as Wanderer does but receives the parameters parsing arguments from the URL. %This allows, for instance, programmatic access in order to perform bulk queries.


% In order to offer batch analysis, data sharing and linking from other Web tools we provide a simplified flavour of Wanderer. This application, Wanderer\_api (\url{http://maplab.cat/wanderer_api}), receives the  

% esulting graphs and tables may be downloaded and an application programming interface (API) allows data sharing, automatable query of multiple instances and direct linkage from external server

\subsection*{Usage examples}
  
Below we report some queries illustrating issues that can be easily resolved by Wanderer, but requiring a large number of steps and/or computational skills to be achieved by other means.

% Some example of issues that can be resolved by querying Wanderer (and not easily reachable without this tool) include:

\begin{itemize}
\item Gene expression profiles. Are different samples expressing the same gene isoforms? Which exon(s) should I analyze to find the largest difference between normal and tumor? As shown in Figure \ref{fig:2}, the balance in the abundance of the different exons) of the PIK3CA tumor suppressor gene is maintained in normal and breast cancers as determined by RNA-seq. On the contrary, colorectal cancers show uneven representation of some exons indicating heterogeneous profiles.
\item DNA methylation profiles. Which region of the CpG island should I analyze to find the largest abnormal DNA methylation in tumor samples? Is it the same for different tumor types? As shown in Figure \ref{fig:3}, the PTGIS CpG islands is represented by five HumanMethylation450 BeadChip probes. A high proportion of colon tumors show hypermethylation of the CpG island but the one that is farther from the TSS (denoted with an asterisk) is already partially methylated in normal tissue and therefore is the less discriminative. On the contrary, this position remains fully unmethylated in most normal prostate tissues but is highly hypermethylated in prostate cancer. PTGIS is downregulated in both colon and prostate cancer.
\item DNA methylation profiles. What is the extent of DNA methylation changes in regard to a a specific gene? EN1 CpG island is frequently hypermethyalted in both colorectal and breast cancer (Figure \ref{fig:4}A). Neverthless when zooming out to the flanking regions (Figure \ref{fig:4}B), the profiles clearly show that hypermethylation affects many neighboring CpG islands in colon cancer but not in breast cancer, where hypermethylation is restricted to the EN1 CpG island (denoted with an asterisk).

\end{itemize}

\section*{Discussion}


Integration and mining of TCGA data is a hot topic in clinical bioinformatics. Given the massive nature of data, some of the tools are primarily focused on data retrieval, such as TCGA-Assembler \cite{zhu2014tcga}, and further annotation and indexing, such as Firehose \cite{firehoseweb}. A second group of analytical platforms are devoted to provide a comprehensive view of the multiple layers of TCGA by linking to the clinical features. An outstanding example is cBioPortal \cite{gao2013integrative}, that delivers, for instance, survival analysis. Regulome explorer \cite{cancer2012comprehensive} shows associations between clinical and molecular features and provides a circular-based visualization of the data, among others. The Cancer Genome Workbench (CGWB) \cite{zhang2010cancer} provides diferents interfaces to integrate and visualize multiple layers of data. Finally, other tools with a broader scope, as the UCSC Cancer Genome Browser \cite{kent2002human}, have incorporated TCGA data allowing genomic coordinate-based display. A comprehensive list of cancer data visualization tools may be found at TCGA website \cite{tcgaanalytical} and recent reviews \cite{plass2013mutations,schroeder2013visualizing}. These ambitious initiatives are invaluable contributions to facilitate the access and analysis of the vast amounts of data but may result deceiving to wet lab researchers and clinicians with limited computational skills and just willing to query very specific information. To face these needs, some tools have been developed providing a very simple interface specialized in a few layers of information. Remarkable examples are MethHC \cite{huang2014methhc}, and MENT \cite{baek2013ment}, mainly focused on the comparison of gene expression and DNA methylation.\\ 

Similarly to these two tools, Wanderer offers a gene-centered access to gene expression and DNA methylation but providing an interactive view in a regional context, which represents a very simple and intuitive representation of otherwise complex information not easily reachable in other packages. Direct visualization of the expression and DNA methylation profiles may be important for the interpretation of the results and to design the most appropriate strategy (choosing the region or the probe to be analyzed) to measure either the expression of the DNA methylation of a given gene in an experimental settings. These scenarios are common in molecular studies and the utility of the information is not limited to cancer research. Basic and clinical researchers working in any field of human biology may be also interested in exploring the features of their favorite gene, especially in normal tissues.\\

Another important feature provided by Wanderer is the simple interface to download small subsets of data in a format easily amenable for customized statistical and graphical analyses. Any user with average proficiency in office suites (i.e: Microsoft Office, OpenOffice, IWorks, etc.) or statistics packages (SPSS, SAS, R, etc.) just by pressing a button can obtain the desired dataset in a compatible format.\\

\section*{Conclusions}

Wanderer is an intuitive interface to explore and interpret gene-associated profiles of expression and DNA methylation for all the cancer types available at TCGA. Normal-tumor paired comparisons are readily provided in the form of graphs and comprehensive tables, facilitating the selection of candidate loci to be considered in experimental and statistical settings. The outputs may be downloaded and shared via API. Wanderer is freely accessible at \url{http://maplab.cat/wanderer}.\\



\section*{Methods}
The user-friendly web application has been developed using R/shiny and a PostgreSQL backend using an eXtreme programming software development methodology. Wanderer is compatible with most common web browsers and operative systems, requires no installation, and can be run on commodity computers, such as low-memory laptops.\\
 
Data consists on HumanMethylation450 BeadChip and Illumina HiSeq exon quantifications as computed by the RNASeq Version 2 pipeline. Data was downloaded with TCGA-Assembler \cite{zhu2014tcga}. Source code is available under the GPL v2 terms at \cite{sourceforge}.







%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                                          %%
%% Backmatter begins here                   %%
%%                                          %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\begin{backmatter}

\section*{Competing interests}
MAP is cofounder and equity holder of Aniling, a biotech company with no interests in this paper. The other authors declare that no competing interests exist.

\section*{Author's contributions}
   ADV, IM and MAP conceived the project. ADV downloaded the data and coded the plots and statistical analysis. IM designed the database and the interface to it. ADV and IM implemented the application and designed the web page. IM and MAP wrote the paper. All the authors read and approved it.

\section*{Acknowledgements}
We thank Iñaki Martinez for excellent technical support. This work was supported by a grant from the Spanish Ministry of Economy and Competiveness (SAF2011/23638). ADV was supported in part by contract PTC2011-1091 from the Spanish Ministry of Economy and Competiveness. The application published here is based upon data generated by the TCGA Research Network: \url{http://cancergenome.nih.gov/}. 


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                  The Bibliography                       %%
%%                                                         %%
%%  Bmc_mathpys.bst  will be used to                       %%
%%  create a .BBL file for submission.                     %%
%%  After submission of the .TEX file,                     %%
%%  you will be prompted to submit your .BBL file.         %%
%%                                                         %%
%%                                                         %%
%%  Note that the displayed Bibliography will not          %%
%%  necessarily be rendered by Latex exactly as specified  %%
%%  in the online Instructions for Authors.                %%
%%                                                         %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% if your bibliography is in bibtex format, use those commands:
\bibliographystyle{bmc-mathphys} % Style BST file



% \bibliographystyle{plain}
\bibliography{wanderer_paper}      % Bibliography file (usually '*.bib' )


% or include bibliography directly:
% \begin{thebibliography}
% \bibitem{b1}
% \end{thebibliography}

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%                               %%
%% Figures                       %%
%%                               %%
%% NB: this is for captions and  %%
%% Titles. All graphics must be  %%
%% submitted separately and NOT  %%
%% included in the Tex document  %%
%%                               %%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%
%% Do not use \listoffigures as most will included as separate files

\section*{Figures}

\begin{figure}[h!]
  \caption{\csentence{Snapshot of Wanderer webpage.}
As an illustrative example, the DNA methylation of HumanMethylation450 BeadChip probes associated to CDKN2A gene in colon cancer are shown. Boxes indicate the different panels for gene name/ID and dataset selection (a), customization of the viewed region (b) and graphs parameters (c), profiles (d) and data download and links (e). 
}
  \label{fig:1}
\end{figure}


\begin{figure}[h!]
  \caption{\csentence{Gene expression profile of PIK3CA in Breast invasive carcinoma (BRCA) and Colon adenocarcinoma (COAD).}
PIK3CA expression levels are consistently reproduced among most exons in normal and tumor tissues in breast cancer patients. In contrast, colon cancers display very heterogeneous profiles with uneven changes among different exons (see enlarged inset for comparison). 
  }
  \label{fig:2}
\end{figure}


\begin{figure}[h!]
  \caption{\csentence{DNA methylation profile of PTGIS CpG island in Colon adenocarcinomas (COAD) and Prostate adenocarcinomas (PRAD).}
Note that cg1077290 probe (*) is already very methylated in colon normal tissue, offering poor discriminant resolution to detect tumor hypermethylation compared with the rest of CpG island probes (labeled in green). In contrast, this probe (cg1077290) shows the highest level of hypermethylation in Prostate cancers (PRAD) and remains unmethylated in normal tissue, resulting as the most discriminant variable in this type of cancer.  
  }
  \label{fig:3}
\end{figure}

\begin{figure}[h!]
  \caption{\csentence{DNA methylation profile of the EN1 gene CpG island and flanking regions in colon (COAD) and breast carcinomas (BRCA).}
(A) Individual and mean methylation levels of EN1 CpG island and adjacent CpGs (chr2: 119,605,000 – 119,606,000). The region remains unmethylated in normal tissues and becomes hypermethylated in both colon and breast cancers. (B) Zooming out the region (chr2:119,540,000-119,706,000) reveals the presence of a large number of CpG islands (green probes) and global hypermethylation of most of them in colon adenocarcinomas. Alternatively breast cancers display interspersed regions of hypomethylation and hypermethylation. The graphs display the average methylation values of each dataset. Probes marked with an asterisk denote statistically significant differences between normal and tumor samples. 
}
  \label{fig:4}
\end{figure}



% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% %%                               %%
% %% Tables                        %%
% %%                               %%
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% %% Use of \listoftables is discouraged.
% %%
% \section*{Tables}
% \begin{table}[h!]
% \caption{Sample table title. This is where the description of the table should go.}
%       \begin{tabular}{cccc}
%         \hline
%            & B1  &B2   & B3\\ \hline
%         A1 & 0.1 & 0.2 & 0.3\\
%         A2 & ... & ..  & .\\
%         A3 & ..  & .   & .\\ \hline
%       \end{tabular}
% \end{table}

% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% %%                               %%
% %% Additional Files              %%
% %%                               %%
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% \section*{Additional Files}
%   \subsection*{Additional file 1 --- Sample additional file title}
%     Additional file descriptions text (including details of how to
%     view the file, if it is in a non-standard format or the file extension).  This might
%     refer to a multi-page table or a figure.

%   \subsection*{Additional file 2 --- Sample additional file title}
%     Additional file descriptions text.


\end{backmatter}
\end{document}
