\documentclass{article}

\usepackage[a4paper, total={6in, 8in}]{geometry}


\title{TCGA Wanderer: integration and visualization of TCGA Data}
\author{Anna D{\'i}ez-Villanueva$^*$, Izaskun Mallona $^*$ and Miguel A. Peinado$^{**}$ \\
\\
 $^*$ Contributed equally to this work\\
 $^{**}$ Corresponding author
}

\date{\today}

\begin{document}

\maketitle

\section{Paper}

Data analysis on human cancer is a hotspot in basic and clinical research. As a group of largely heterogenous and complex disease, cancer research is benefited by coordinated efforts on data acquisition and analysis. The Cancer Genome Atlas (TCGA) is a result of an international effort that offers a multilayered view of the genomics and epigenomics of more than 30 human cancer types \cite{weinstein2013cancer}. This enormous data availability opens a wide world of possibilities to the research community. For instance, it allows \textit{in silico} hypothesis testing or validation of in-house results through an independent series.\\

The TCGA serves the information in three tiers according to the degree of its bioinformatical processing: raw, lightly and heavily analyzed data (named levels 1, 2 and 3). Data not belonging uniquely to an individual, such as level 3 and de-identified clinical data, is open access; on the other hand, the download of protected data, such as raw sequencing files, requires authentication via the The Cancer Genomics Hub (CGHub)\cite{wilks2014cancer}. Finally, although the open access data tier offers immediate access to the information, some projects restrict the dissemination of the results brought by its use until a first global analysis is published by the consortia.\\

The TCGA Data Portal is the gateway to retrieve the TCGA information. Basically, offers a text file for each processed sample belonging to a dataset subjected to a platform; compressed packages for multiple samples are allowed. For instance, it allows downloading 300 files with the clinical data related to samples of the Acute Myeloid Leukemia (LAML) dataset, but only provides RNAseq results for roughly a half of them. It is worth noting that as the number of files and their sizes increase, further processing of this information exceeds the spredsheet-based analysis capabilities and requires some amount of bioinformatics knowledge, (i.e. programming under the R statistical environment). Comprehensive and integrative analysis is indeed dependent on informatics solutions, such as TCGA Assembler \cite{zhu2014tcga} or the cBioPortal\cite{gao2013integrative}.\\

The availability of such a large amount of data is exciting to the cancer research community as it helps to check candidate \textit{loci}. For instance, as some genes show aberrant promoter hypermethylation and therefore silencing during the tumor progression \cite{jones2002fundamental}, experimentalists might query the expression and methylation profiles of their candidate \textit{loci} on tumoral and normal matched tissues. However, the Data Portal does not allow to query individual genomic locations but rather offers bulky, whole genome datasets. To our knowledge, although much effort has been done on facilitating the access to TCGA data, none of the published tools allows an user-friendly interface to solve this kind of queries. We believe that this candidate approach might be particularly benefited by using RNASeq and methylation data.\\


With this aim we developed a tool, the TCGA Wanderer, that allows to easily retrieve and visualize the regional profiles of groups of samples belonging to a tumor type. The user is only asked to introduce the TCGA dataset (i.e. colon adenocarcinoma), whether the query is related to methylation or expression, and the \textit{locus}; gene symbol, Ensembl ID, exon or methylation probenames are valid target identifiers. The result layout summarizes in a upper panel the tumor's profile and the lower, the matched normal's. Each profile links by lines the selected values by sample, meaning that sample heterogeneity may be easily inspected by eye. Gene location and sense and CpG islands (for methylation plots) are also printed. The plots are highly dynamic, allowing to move through a region and to zoom in and out. Images and the TCGA's data they represent are downloadable just by clicking a button. Currently the datasets available are: colon adenocarcinoma (COAD), breast invasive carcinoma (BRCA), lung adenocarcinoma (LUAD) and thyroid carcinoma (THCA).  \\

% talk about the human methylation340k array

The user-friendly web application has been developed using R/shiny and is freely accesable at xx. As it is hosted in a shiny server, the final user only needs a web browser to run it and requires no installation, being platform-indepedent and runnable on commodity computers, such as low-memory laptops.\\

\subsection{Acknowledgements}
The application published here is based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/.



\bibliographystyle{plain}
\bibliography{wanderer_paper}      % Bibliography file (usually '*.bib' )




\section*{Cover letter}


Dear Editor,\\

We would like to submit the manuscript entitled \textit{TCGA Wanderer: integration and visualization of TCGA Data} for consideration as a Correspondence letter to Nature Methods.\\

The Cancer Genome Atlas (TCGA) iniciative is an exciting oportunity to check comprehensively the genomic and epigenomic abnormalities on clinically characterized cancersamples. However, the easy access to this data is far away from the wet lab experimentalists, as the amount of data offered is very large.\\

Particularly, analysis of DNA methylation is probably one of the hottest issues in biomedical research: in 2013 about 900 papers included this term in the title and about 4000 in the abstract, with an annual growth rate >20\% during the last 5 years (data from PubMed). The abnormalities on gene expression and its coupling to DNA methylation during tumorogenesis are striking features that have been atracting researches attention since the 2000.\\

Here we describe a new Web-tool that provides an easy-to-use interface to fetch and visualize expression and methylation data from the TCGA. The user is only asked to introduce the tumor type (i.e. prostate cancer), whether the query is related to methylation or expression, and the gene symbol. Then, the tool retrieves and plots the results in a highly reactive fashion, providing a dynamic image rendering that allows the generation of publication-ready plots. The outputs are easy to download, including both the plots and the spreadsheets.\\

To our knowledge this is the first tool that offers this capabilities. We are convinced it will be of great interest for a broad spectrum Nature Methods readers.\\


Sincerely, \\
Izaskun Mallona, PhD.

\end{document}