\documentclass{article}

\usepackage[a4paper, total={6in, 8in}]{geometry}


\title{TCGA Wanderer: integration and visualization of TCGA Data}
\author{Anna D{\'i}ez-Villanueva$^*$, Izaskun Mallona $^*$ and Miguel A. Peinado$^{**}$ \\
\vspace{3cm}
 $^*$ Contributed equally to this work
 $^{**}$ Corresponding author
}

\date{\today}

\begin{document}

\maketitle

\section{Correspondence}

Data analysis on human cancer is a hotspot in basic and clinical research. A group of largely heterogenous and complex disease, cancer research is benefited by coordinated efforts on data acquisition and analysis. The Cancer Genome Atlas (TCGA) is a result of an international effort that offers a multilayered view of the genomics and epigenomics of more than 30 human cancer types \cite{weinstein2013cancer}.\\

This enormous data availability opens a wide world of possibilities to the research community. For instance, it allows \textit{in silico} hypothesis testing or validation of in-house results through an independent series. With this aim, TCGA Data Portal serves the information in three tiers according to the degree of the bioinformatical processing: raw, lightly and heavily analyzed data (named levels 1, 2 and 3). Data not belonging uniquely to an individual, such as level 3 and de-identified clinical data, is open access; on the other hand, the download of raw sequencing data requires authentication via the The Cancer Genomics Hub (CGHub)\cite{wilks2014cancer}.\\ 

Although the open access data tier offers immediate access to the information, some projects restrict the dissemination of the results brought by its use until a first global analysis is published by the consortia.\\

Open access data is available through the TCGA Data Portal\cite{zhang2011international}, that permits to download text files containing processed data for a set of tumors. However, as the data size reaches XX rows, further processing of this information exceeds the spredsheet-based analysis capabilities and requires some amount of bioinformatics knowledge, (i.e. programming under the R statistical environment). Comprehensive and integrative analysis is indeed dependent on informatics solutions, such as TCGA Assembler\cite{zhu2014tcga} or the cBioPortal\cite{gao2013integrative}.\\

The availability of such an amount of data may aid to check the behaviour of candidate \textit{loci}. As some genes show aberrant promoter hypermethylation and therefore silencing during the tumor progression \cite{jones2002fundamental}, experimentalists might query the expression and methylation profiles of their candidate \textit{loci} on tumoral and normal matched tissues. To our knowledge, although much effort has been done on facilitating the acces to TCGA data, none of the published tools allows an user-friendly interface to solve this kind of queries.\\

%
We developed a tool, the TCGA Wanderer, that allows to easily retrieve and visualize the regional profiles of groups of samples belonging to a tumor type. The user is only asked to introduce the TCGA dataset (i.e. colon adenocarcinoma), whether the query is related to methylation or expression, and the \textit{locus}; gene symbol, Ensembl ID, exon or methylation probe identifiers are valid target identifiers. The result layout summarizes in a upper panel the tumor's profile and the lower, the matched normal's. Each profile links by lines the selected values by sample, meaning that sample heterogeneity may be easily inspected by eye. Gene location and sense and CpG islands (for methylation plots) are also printed. The plots are highly dynamic, allowing to move through a region and to zoom in and out. Images and the TCGA's data they represent are downloadable just by clicking a button. Currently the datasets available are: colon adenocarcinoma (COAD), breast invasive carcinoma (BRCA), lung adenocarcinoma (LUAD) and thyroid carcinoma (THCA).  \\

% talk about the human methylation340k array

The user-friendly web application has been developed using R/shiny and is freely acesable at xx. As it is hosted in a shiny server, a web browser is needed to run it and requires no installation, being platform-indepedent and runnable on commodity computers, such as low-memory laptops.\\

\subsection{Acknowledgements}
The application published here is based upon data generated by the TCGA Research Network: http://cancergenome.nih.gov/.



\bibliographystyle{plain}
\bibliography{wanderer_paper}      % Bibliography file (usually '*.bib' )




\section{Letter to editor}



\end{document}